### DISCUSSION TITLE
Идеи и предложения

### DISCUSSION THREAD

**Discussion by vvip-68** [vvip-68/GyverPanelWiFi#196] [Идеи]
Придумали что-то новое? 
Предлагайте свои идеи тут.

**Comment by kostyamat** [comment]
Стандартизироваться в мапинге E1.31 с прошивкой FireLamp_JeeUI, а так же можете забрать себе трансляцию из Soulmate IDE, если желание есть.
https://community.alexgyver.ru/threads/wifi-lampa-budilnik-proshivka-firelamp_jeeui-gpl.2739/post-103219

https://community.alexgyver.ru/threads/wifi-lampa-budilnik-proshivka-firelamp_jeeui-gpl.2739/post-107099

Взять трансляции можно тут https://github.com/DmytroKorniienko/FireLamp_JeeUI/blob/dev/include/ledStream.h

https://github.com/DmytroKorniienko/FireLamp_JeeUI/blob/dev/src/ledStream.cpp

**Reply by vvip-68** [reply]
Да переделать наверное можно... А в чем смысл? Совместимости между проектами все равно нет никакой.
Вот я знаю уже нескольких человек сделавших настройки под этот проект. Да, в вашем случае проще мапить целыми рядами. Но им же все переделывать придется.

За пример с SoulMate - спасибо. Рассмотрю поподробнее...

**Comment by alexder21** [comment]
1, Добавить управление яркостью в макросы и/или по расписанию.
2, Перенести управление из приложения на www (если хватит памяти)

**Reply by simonovich** [reply]
Наверное имеется ввиду добавить возможность управления через http запросы.
Если хочется пощупать "веб-морду", можно прикрутить какой-нибудь фронт-клиент аля "умный дом", который послужит прослойкой между вами и mqtt. В этом случае никаких доработок не требуется.

**Reply by alexder21** [reply]
Я имел ввиду, чтобы не "тащить" телефонное приложение, заменить на "веб-морду".

**Reply by Forestman99** [reply]
согласен, яркость по расписанию бы не помешала, сейчас на ночь только часы в ночном режиме пустить можно, а уменьшить яркость чтобы с балкона не сильно светило точно надо, насколько это сложно?

**Reply by vvip-68** [reply]
В основном - переписать приложение, добавив туда возможности задавать период, когда яркость должна меняться и установку значения этой яркости. Исходный код приложения лежит в папке Android проекта. Попробуйте загрузить его в AppInventor и что-нибудь поделать. Поймете насколько это мучительно.

В самой же прошивке - гораздо меньше работы и изменений. Но без доделки приложения - идея настраивать время и яркость перепрошивкой контроллера - так себе идея...

**Reply by Forestman99** [reply]
> В основном - переписать приложение, добавив туда возможности задавать период, когда яркость должна меняться и установку значения этой яркости. Исходный код приложения лежит в папке Android проекта. Попробуйте загрузить его в AppInventor и что-нибудь поделать. Поймете насколько это мучительно.
> 
> В самой же прошивке - гораздо меньше работы и изменений. Но без доделки приложения - идея настраивать время и яркость перепрошивкой контроллера - так себе идея...

не, пойти путем проще
в приложении в расписании добавить два пункта - яркость минимум и яркость дефолт(что была), в прошивке вместо эффекта в указанное время менять яркость на минимальную и по умолчанию(запомненную ранее)

**Reply by Forestman99** [reply]
> Поймете насколько это мучительно.

а вот это перемещение рывками и тормоза по минуте в Блоках АпИнвертора - это норма или у меня инет/пк тормозят?

**Reply by vvip-68** [reply]
Это норма для больших проектов. А это приложение - большой проект. И, да - зависит от скорости соединения с инетом тоже. Там на каждый "чих" передается/принимается довольно большой объем. Скорее всего - это картинка области отображения и еще какие-то метаданные...

Потому и говорю - добавлять что-то новое в это приложение - мука мучительная.
Вы это, похоже, уже прочувствовали.

**Reply by Forestman99** [reply]
> Потому и говорю - добавлять что-то новое в это приложение - мука мучительная. Вы это, похоже, уже прочувствовали.

для минимизации муки - вот в список эффектов добавить два пункта, минимальная яркость и дефолтовая,
![Screenshot_5](https://user-images.githubusercontent.com/34151320/154934034-ed8954b1-355a-4271-9371-f688d17cb6b0.jpg)
тут индекс прибавить 6 вместо 4
![Screenshot_4](https://user-images.githubusercontent.com/34151320/154934048-2d9ab8ad-ecdd-4454-bcaf-2766fc27f554.jpg)
в прошивке - 
количество эффектов увеличить на 2, дать им ID и в обработке режимов прописать вызов яркости с нужными параметрами.
Нет опыта в AI совсем, не понимаю где там массив со списком эффектов подправить

**Reply by vvip-68** [reply]
Список эффектов передается из прошивки в приложение.
При получении он предваряется постоянными значениями.
Происходит это примерно тут:

![image](https://user-images.githubusercontent.com/39353139/154955034-361fdcd1-0ed0-42ac-a7cc-8236099c0999.png)

**Reply by Forestman99** [reply]
> Список эффектов передается из прошивки в приложение. При получении он предваряется постоянными значениями. 

то есть для моего варианта менять только прошивку, вроде в прошивке нашел место где выбирается эффект и смотрится время и указания на него

**Reply by Forestman99** [reply]
так, добавил в список ID две пункта с яркостью, в массив названий добавил, в куске кода где действия при начале/окончании другие режимы добавил смену яркости через fastled, после отработки выключается авторежим, буду курить логику и разбирать код дальше

**Reply by Forestman99** [reply]
я не суперкодер, сделал так что при выборе режима происходит изменения яркости и включается рандомный режим - к моим требованиям подходит

1. a_def_soft.h добавлены эффекты Камин,Стрелки, **Яркость MIN, Яркость MAX**, Погода,Жизнь
2. a_def_soft.h добавлены в список **ID**
 ```
#define MC_BRIGHT_MIN           38
#define MC_BRIGHT_MAX           39
 ```
3. a_def_soft.h ниже увеличены ID для последующих эффектов с SD карты
4. custom - в процедуре **processEffect** добавлены действия
```
case MC_BRIGHT_MIN:          globalBrightness=25;FastLED.setBrightness(globalBrightness); setManualModeTo(false); setRandomMode(); break;
case MC_BRIGHT_MAX:          globalBrightness=230;FastLED.setBrightness(globalBrightness); setManualModeTo(false); setRandomMode(); break;
``` 

5. custom - в процедуре **releaseEffectResources** добавлены действия
```
case MC_BRIGHT_MIN:          break;
case MC_BRIGHT_MAX:          break;
```
в итоге выглядит вот так расписание, в указанное время меняется яркость и идет рандомный разрешенный эффект
![Screenshot_7](https://user-images.githubusercontent.com/34151320/157283155-14bc7162-4461-4c5d-91dc-bd00f06bfeb9.jpg)

**Comment by kost111233** [comment]
1.Добавить в приложение возможность менять последовательность эффектов. 
2.Добавить возможность после какого-либо макроса, например после бегущей строки с текстом «С НОВЫМ ГОДОМ» запустить определенный эффект, у меня для этого уже подготовлен эффект салют:)

3. хотелось бы такое же приложение на айфон, но уже видел писали об этом, а заниматься никто не хочет и скорей всего не будет

**Reply by vvip-68** [reply]
2. В тексте строки можно задать на фоне какого эффекта он должен показываться - пусть бежит на фоне эффекта "Салют".
3. Да, никто не хочет. Тот кто имеет айфоны - не умеет программировать, те, кто умеют программировать - не любят яблочные девайсы...

**Reply by kost111233** [reply]
> * В тексте строки можно задать на фоне какого эффекта он должен показываться - пусть бежит на фоне эффекта "Салют".
> * Да, никто не хочет. Тот кто имеет айфоны - не умеет программировать, те, кто умеют программировать - не любят яблочные девайсы...

Про фон знаю и пользуюсь, спасибо. Не на всех эффектах хорошо читается текст. Возможно эта задумка с эффектом без текста только мне интересна:)
 
Еще вопрос, если у меня на sd карте больше одного ролика, можно в макросе прописать  текст на фоне определенного эффекта на флешке? В приложении можно выбрать просто SD

Жаль конечно, что никто не хочет с яблоками связываться, но и без этого все равно очень крутой проект, спасибо

**Reply by vvip-68** [reply]
> Не на всех эффектах хорошо читается текст. 
Для этого у эффекта есть параметр "контрастность" - ее нужно установить таким образом, чтобы яркость эффекта была меньше яркости текста и текст хорошо читался на фоне эффекта.

> Еще вопрос, если у меня на sd карте больше одного ролика, можно в макросе прописать текст на фоне определенного эффекта на флешке? 

Да, что-то такое было. Насколько я помню фон в качестве эффекта - {Exx} - где xx - номер эффекта. Если нужно выбрать конкретный вариант эффекта (читай - ролик на карте для SD) - нужно использовать макрос {Exx-n} - тут n - номер варианта. Нужно учитывать, что 0 - это "Случайный выбор", 1 - это "Последовательно" и только 2 - это первый файл в списке эффектов.

Все о макросах - [тут](https://github.com/vvip-68/GyverPanelWiFi/wiki/Настройка-бегущей-строки)

**Comment by simonovich** [comment]
WiFi Panel Setup - можно подключиться по MQTT
WiFi Panel - можно подключиться по MQTT
WiFi Panel Player - нельзя... :(

**Reply by vvip-68** [reply]
Смысла никакого нет. Player - интерактивное приложение.
Игры - задержки по MQTT такие, что реакция на действие в игре придет заведомо поздно, когда змейка врежется / фигурка тетриса упадет / в арканоиде шарик уже давно улетит.
Рисование - тоже медленно плюс какой смысл рисовать издалека, когда экрана не видишь, а если экран видишь - зачем тебе MQTT - пользуйся локалкой.

**Reply by simonovich** [reply]
Я себе накидал кнопок в MQTT клиент и гоняю на балконе с лавочки внизу. Оно как бэ изначально для того и задумывалось, правда у Гайвера чуток по-другому.
Змейка - да, сложно. Тетрис - шансы выше. Лабиринт - ваще изи, там скорость не важна.

**Reply by simonovich** [reply]
Если задержки и лаги прям ппц, то единственный выход - прокинуть себе тунель в домашнюю сеть. Но для этого она (сеть) должна торчать наружу с белым ip.
Ест-но, в этом случае MQTT ваще не нужен, можно запускать приложеньку как обычно - все будет как "дома".

**Comment by dina16121974** [comment]
Для есп 32 ,энкодер бы добввить, как в лампе KDNа , было бы здорово.

**Comment by sepulka123** [comment]
Хорошо бы в "JinxFramer"  добавить работу с составными матрицами.

**Comment by sepulka123** [comment]
На сайте "https://editor.soulmatelights.com/gallery" есть много очень симпатичных эффектов.
Хорошо бы иметь возможность копировать эти эффекты в Ваш проект.

**Reply by simonovich** [reply]
А в чём проблема? Там же код к каждому эффекту приложен. И библиотека та же.
Должен быть минимум правок. Только насчет памяти не уверен. Возможно надо будет пересчитывать, чтобы новый эффект воткнуть.

**Reply by sepulka123** [reply]
Ключевое слово  "... правок".  И мне, как не программисту, сложно
определить - что и как править.

**Comment by Ko1hozer** [comment]
Доброго времени суток. Хотелось бы иметь возможность тянуть со своего MQTT данные к примеру с собственной метеостанции. А еще при возможность бегущей строки в несколько строк.

**Comment by Rerol911** [comment]
А как на счет трансляции актуального курса доллара бегущей строки, как это сделать?

**Reply by vvip-68** [reply]
1. Найти сервис, который по запросу отдает текущий курс доллара.
2. По аналогии с сервисом получения погоды написать отправку запроса этому сервису, получение и парсинг данных.
3. Добавить в обработчик бегущей строки какой-нибудь макрос, отображающий текущий курс выбранных валют.

**Reply by vvip-68** [reply]
Курс доллара станет неактуальным в ближайшие, думаю, полгода-год. Не будет у нас хождения доллара, он останется актуальным только для Америки как региональная валюта. Поэтому большой вопрос - стоит ли на это тратить время...

**Reply by Rerol911** [reply]
> 

вообще ни одного слова не понял, но спасибо за помощь

**Reply by simonovich** [reply]
У меня рабочая ось - линукс. Могу быстро написать скрипт, который парсит текущий курс с какого-нибудь сайта и отправляет в порт на матрицу в качестве текста стандартного эффекта бегущей строки.
Под Windows сложнее, но в принципе тоже рабочий вариант.
В чем минус такого варианта: нужен включенный компьютер или ноутбук, на котором будет работать скрипт и отсылать данные на матрицу.

Вариант 2. Чуть похитрее. Включаем на матрице управление по MQTT. Пишем скрипт как для варианта 1, но запускаем его где-нибудь у себя на роутере, или на сервере в интернет. Работать такой скрипт должен через программу MQTT-клиент. Т.е. если у вас нет возможности установить MQTT-клиент на роутер или сервер, то не прокатит. Это минус.
А плюс - не обязательно держать включенным комп, все будет крутиться само.

Вариант 3. Заставить матрицу самостоятельно вытягивать и показывать курсы, без всяких там компьютеров и роутеров.
Для этого придется дописать скетч и перепрошить матрицу.

Это все варианты минимум, т.е. про мобильное приложение молчим, не трогаем.

Нужно ли это в перспективе? Кому как. У нас аренда жилья, кредиты, товары с того же али - все привязано к баксу. Возможно в РФ это не так критично.

Ну, если бакс уйдёт, то можно настроить курс юаня! ))

**Reply by Forestman99** [reply]
в инете куча сервисов выдающих через json курсы разных валют.
ну в проекте есть пример как json ответ парсить, можно вместо сервиса одной из погод заменить и подредактировать как выводить через макросы. саму строку формировать в проекте

**Comment by timof123** [comment]
Предлагаю идею. В общем, сделать возможность отсчета времени до какого-нибудь события, чтобы, например, за минуту начался отсчет в секундах. Да, в прошивке уже есть возможность делать строку с отсчетом, но там просто бежит строка "59 секунд", "58 секунд" "57 секунд" и т.д. Хочется, чтобы это была не бегущая строка, а статичные цифы посередине экрана без приписки "секунд".

**Comment by san-q66** [comment]
Здравствуйте! А никто не заморачивался сделать интеграцию в Home Assistant?
Что бы можно было выводить информацию с сервера текстом, и управлять матрицей.

**Reply by simonovich** [reply]
mqtt есть
можно самому интегрировать

**Reply by RomanLesovoi** [reply]
есть рабочие примеры?
пока что смог только через udp вкл-выкл и статус получить
`command_line:
  - switch:
      name: гирлянда
      command_on: "echo '$1 1;' | nc -w 2 -u 192.168.200.19 2390"
      command_off: "echo '$1 0;' | nc -w 2 -u 192.168.200.19 2390"
      command_state: "echo '$6 7|PS;' | nc -w 1 -u 192.168.200.19 2390 | grep -q 'PS:1;' && echo 1 || echo 0"
      value_template: '{{ value=="1" }}'`

**Reply by simonovich** [reply]
Вот пример коммуникации через serial + mqtt, если вам так проще:
https://github.com/simonovich/usb_button/blob/master/scripts/usb_button.sh

Но я бы выкинул serial из этой цепочки и использовал какие-нибудь mqtt коннекторы из набора вашего HA, что-то вроде этого:
https://cyan-automation.medium.com/setting-up-mqtt-and-mosquitto-in-home-assistant-20eb810a91e6

Примеры здесь не для гирлянды, а просто чтобы показать принцип работы.

**Comment by persivin** [comment]
> Придумали что-то новое? Предлагайте свои идеи тут.
1, Почему в приложении в настройке таймеров нет  действия "Включить матрицу"? Есть "Выключить матрицу" и включить её можно только на определённом эффекте. Но у меня матрица работает в режиме автоматической смены эффектов. Вечером включается , приходится входить в приложение ставить автоматическую смену эффектов и поздно ночью она выключается. Если бы была функция просто включить, она бы сразу включалась с автоматической сменой эффектов.
2, Можно добавить интеграцию в умный дом Яндекса или Tuya , чтобы голосом можно было включать или выключать?

**Reply by vvip-68** [reply]
1.Включить матрицу можно или на указанном эффекте, или в "Демо режим" - тогда она включается с перебором всех эффектов в автоматическом режиме.
2.Наверное можно. Но нет специалистов, которые бы понимали как это делать и что для этого нужно.

**Comment by mbb6111** [comment]
Всех с наступившим Новым 2024 годом!

Может мое предложение и не ново.
Я для питания wi-fi панелей использую блоки питания ATX от компьютеров.
Подключаю так:
![WiFi_Panel_ATX](https://github.com/vvip-68/GyverPanelWiFi/assets/155435406/9f044046-a56f-463b-935e-e4a4e1b76f56)

Если в блоке питания используется микросхема UC3843 или ее аналоги, то в таких блоках есть подстроечный резистор для подстройки выходного напряжения и "+5В" можно уменьшить до "4,8В" без срабатывания защиты (проверено на блоках IP-P300AJ2-0 и IP-S200DF1-0)

**Reply by simonovich** [reply]
Отличная идея! Тоже хотел использовать, но побоялся, что не хватит мощности. Компьютерные БП выглядят серьёзно: бывает 300-400W и больше, но в них эта мощность распределяется на несколько линий - это 3.3V, 5V, 12V и т.д. И конкретно на 5V получается не так уж много от тех 300-400W.
В этом году, точнее уже в прошлом, мои адресные светодиоды были далеко от меня, даже не видел забитых туда поздравлялок... Но аптайм что-то около года, вместе с полным набором всех "перделок", типо часов, музыки, SD карты и т.д. Прошивку не менял, она еще вот с этого самого -  "старого" репозитария. Исправно "бомбит" через mqtt, всегда на связи.
Спасибо за хороший и стабильный проект! Всех с Новым Годом и Рождеством!

**Comment by Cyxapuk11** [comment]
Проект замечательный. Хотелось бы еще в приложении во вкладке "режимы по времени" ползунки с яркость добавить под каждой строкой. Или доработать проект автояркостью по освещенности (датчик в матрицу) или времени суток. А то на маленькой яркости днем слишком темно и эффекты выглядят хреново, а на большой яркости вечером слишком резко бьет по глазам и отвлекает внимание

**Reply by vvip-68** [reply]
1. Конкретно этот проект устарел, закрыт и дорабатываться не будет. Особенно приложения.
2. Автоматическая яркость показала свою полную несостоятельность. Реальные сценарии (использование в реальной жизни) показало несостоятельность реализации линейной зависимости - стемнело - убавили яркость, стало светлее - прибавили яркость. В жизни все гораздо разнообразнее. К тому же подумайте как вы будете компенсировать яркость от гирлянды - влияние на датчик освещенности. В общем - было уже, выкинуто за несостоятельностью.
3. Яркость по времени (или по режимам) - надо подумать как это правильно и непротиворечиво организовать. Возможно в следующей версии проекта что-то такое и будет. Запланировано, но не в ближайшем приоритете.

**Comment by Cyxapuk11** [comment]
Для гирлянды датчик можно направить в другую сторону. В сборной же матрице вообще не вижу проблемы, он будет стационарно расположен так, чтобы первоочередно учитывать окружающую освещенность. 
На крайняк костыль: раз прошивка умеет предугадывать ток потребления, значит она понимает и текущую яркость матрицы. Сделать параметр, который будет вычитать яркость матрицы из показаний датчика с каким-то множителем (ну или там показательные/логарифмические всякие действия, не очень шарю что тут лучше применимо).

Понимаю и принимаю ваш отказ, просто люблю подискутировать

**Comment by shkolin1973** [comment]
Здравствуйте!
Возникла такая сложность. Сделал матрицу на балконе 34х17. 4 стекла - 4 секции матрицы. Высота 17 точек. Ширина 8-9-8-9 столбцов. Итого: 34 столбца. Балкон Г-образный, т.е. последняя секция (9 столбцов) под 90град к остальным. 
При выводе часов маленьким шрифтом, всё ок, часы полностью помещаются на прямом участке.
Но хочется часы с большим шрифтом. А там минуты и знак градусов температуры залазят на 4-ю секцию (которая под 90град).
Ничего не понимаю в программировании ардуино, но почитав описания в скетче относящиеся к часам, понял, что там всё центруется относительно середины матрицы.
Можно ли часы сдвинуть в лево на первые 8-9-8 столбцов. Можно-ли что-то поменять в скетче? Если можно подскажите, пожалуйста, где? В идеале сделать смещение часов настраиваемым параметром (это так, в качестве предложения).

**Reply by vvip-68** [reply]
Ну, параметр особенно в версии 1.13 я точно делать не буду.
Для вашего случая в файле clock.ino добавьте следующее:

![image](https://github.com/user-attachments/assets/0cce2cf0-ce07-4d9d-89d8-1afb6e4f5171)

![image](https://github.com/user-attachments/assets/e48dc7da-4e45-4f09-9c71-cf2172a65a85)

-5 - смещение на 5 точек влево. Если +5 - то в право на 5 точек.
Поставьте ваши значения.

**Reply by shkolin1973** [reply]
Спасибо за ответ! 
немного не понятно...
Я понял, что правку нужно делать в двух местах, но в примере у вас в 307 строке  5 (+5), а в 797 стоке  -5.
Т.е. и в 307 и 797 строках должно быть или плюс или минус (столько, сколько мне нужно)

**Reply by vvip-68** [reply]
напишите в 307 строке так
x += -5

Чтобы было одинаково

**Reply by shkolin1973** [reply]
Понял, Спасибо:-)))
Буду пробовать!

**Reply by shkolin1973** [reply]
Добрый вечер! Спасибо за помощь! 
Почти получилось! Сместил на -4. Но есть одно НО. Фактически та ширина матрицы на которую нужно вместить часы 25 столбцов. А  матрица на самом деле шире, и как я понял поэтому выводятся двойные точки и пробелы до точек и после. Короче когда в показаниях времени нет единицы (все цифры занимают по 5 столбцов) крайний правый столбец "заезжает" на боковую часть матрицы. Пытался сам понять, но так и не понял: как убрать пробел рядом с точками или как точки сделать одинарными. 
Прошу ещё раз помочь. Если получится "ужать" большие часы на один столбец будет вообще идеально! Кстати и малые часы отцентровались и тоже смотрятся лучше.

**Reply by vvip-68** [reply]
Ну попробуйте еще вот эт
![image](https://github.com/user-attachments/assets/865ec00b-1e05-42a9-b7c8-ffe86879a387)

Зачеркнутое удалить. Внизу там где выделено - все строку удалить вместо удаленной строки   написать то что красным текстом - x += 1;

**Reply by shkolin1973** [reply]
Спасибо большое! Всё получилось! Часы "ужались" на 2 столбца. (-1 пробел и точки стали в один столбец). То, что нужно!

**Reply by shkolin1973** [reply]
Окончательно идеальное смещение для меня оказалось -3. При -4  и когда температура состоит из двух "широких" цифр (-20 например) знак минуса состоял из 2-х точек, а справа столбец был свободен. При -3 всё везде влазит. Спасибо за помощ!
Может кому ещё пригодится.

**Comment by shkolin1973** [comment]
Здравствуйте! 
А вы не думали добавить получение данных о погоде с собственного датчика? В идеале с BME280 или хотя бы с DS18B20.
Или для меня вообще бы идеально было получение данных о температуре/влажности/давлении с сервиса thingspeak.
У меня метеостанция по этому проекту https://radiokot.ru/circuit/digital/home/245/ туда (thingspeak) передает данные. А другое устройство собранное по этому же проекту, но с экраном забирает данные с thingspeak и отображает на дисплее. Удобно, сами метео датчики на улице, контроллер на подоконнике (на балконе), а контроллер с дисплеем в любом месте в доме. И между ними нет проводов.  И ещё бы в гирлянду получать эти данные и отображать реальную погоду за окном. Если сложно, то и не надо. Чаще всего разница реальной температуры и той, что приходит из Яндекса не большая, 1-2 градуса, но иногда и до 6-8 доходит.
